framework/src/media/audio/audio_manager.c

```diff
@@ -576,20 +577,25 @@ audio_manager_result_t set_audio_stream_in(unsigned int channels, unsigned int s
 		g_audio_in_cards[g_actual_audio_in_card_id].resample.buffer = malloc((int)((float)get_input_frames_to_byte(get_input_frame_count()) / g_audio_in_cards[g_actual_audio_in_card_id].resample.ratio) + 1);	// +1 for floating point margin
 		if (!g_audio_in_cards[g_actual_audio_in_card_id].resample.buffer) {
 			meddbg("malloc for a resampling buffer(stream_in) is failed\n");
-			pcm_close(g_audio_in_cards[g_actual_audio_in_card_id].pcm);
-			return AUDIO_MANAGER_RESAMPLE_FAIL;
+			ret = AUDIO_MANAGER_RESAMPLE_FAIL;
+			goto error_with_pcm;
 		}
 	}
 
 	pthread_mutex_unlock(&(g_audio_in_cards[g_actual_audio_in_card_id].card_mutex));
-
-	return AUDIO_MANAGER_SUCCESS;
+	return ret;
+	
+error_with_pcm:
+	pcm_close(g_audio_in_cards[g_actual_audio_in_card_id].pcm);
+	pthread_mutex_unlock(&(g_audio_in_cards[g_actual_audio_in_card_id].card_mutex));
```

Comment:
I see that pthread_mutex_lock/ unlock has been performed to protect the global resource  g_audio_in_cards , But i have a concern here, What if the context switch happens after the line no 540 and then somebody modifies the g_actual_audio_in_card_id to INVALID as that global variable is not protected ?  
Isn't it the rest of the logic after line no 40 leads to wrong result ?  What is your opinion?

Reply:
yes I understand what you concerning, But actually g_actual_audio_in_card_id is kindly temp code, will fix later.